---
title: "TCGA LUAD"
author: "Kindle Zhang qz2527"
date: "2024-12-26"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      message = FALSE,
                      warning = FALSE)
library(tidyverse)
library(BiocManager)
library(TCGAbiolinks)
```

# get gene expression form

## download data

```{r download data}
# setwd('data_file')
# cancer_type = "TCGA-LUAD"
# # TCGA abbreviation：https://www.jianshu.com/p/3c0f74e85825
# # ready to download 
# expquery = GDCquery(
#   project = cancer_type,
#   data.category = "Transcriptome Profiling",
#   data.type = "Gene Expression Quantification",
#   workflow.type = "STAR - Counts"
# )
# # start to download in working place
# GDCdownload(expquery, directory = "GDCdata")
# # transform
# expquery2 = GDCprepare(expquery,
#                        directory = "GDCdata",
#                        summarizedExperiment = T)
# # save rda
# save(expquery2, file = "luad.gdc_2022.rda")
```

## import data

```{r}
load("data_file/luad.gdc_2022.rda") # import
load("data_file/gene_annotation_2022.rda") # annotation
table(gene_annotation_2022$type) 
```

the name of gene is ENSEMBL and symbol. Then we get the data table gene name vs sample

```{r counts}
counts = expquery2@assays@data@listData[["unstranded"]] # integer
colnames(counts) = expquery2@colData@rownames
rownames(counts) = expquery2@rowRanges@ranges@NAMES

# transfer row names to gene name
counts = counts |>
  as.data.frame() |>
  rownames_to_column(var = "ENSEMBL") |>
  inner_join(gene_annotation_2022, "ENSEMBL") |>
  # .[!duplicated(.$symbol), ]
  distinct(symbol, .keep_all = TRUE)

rownames(counts) = NULL
counts = counts |>  column_to_rownames("symbol")

# we focus on the mRNA so we use protein_coding, maybe lncRNA next time
table(counts$type)
counts = counts[counts$type == "protein_coding", ]

# simplify table
counts = counts[,-c(1, ncol(counts))]
ncol(counts)
nrow(counts)

# simplify name
colnames(counts) = substring(colnames(counts), 1, 16)
counts = counts[,!duplicated(colnames(counts))]

# sample type
table(substring(colnames(counts), 14, 16))

# normal sample
# retain 01A
counts01A =
  counts[, substring(colnames(counts), 14, 16) == c("01A")]
# cancer sample
# retain 11A
counts11A =
  counts[, substring(colnames(counts), 14, 16) == c("11A")]
table(substring(colnames(counts01A), 14, 16))
table(substring(colnames(counts11A), 14, 16))
```

The counts only be used to make a difference analysis

```{r tpms}
tpms = expquery2@assays@data@listData[["tpm_unstrand"]]
colnames(tpms) = expquery2@colData@rownames
rownames(tpms) = expquery2@rowRanges@ranges@NAMES

tpms = tpms |>
  as.data.frame() |>
  rownames_to_column("ENSEMBL") |>
  inner_join(gene_annotation_2022, "ENSEMBL")
tpms = tpms[!duplicated(tpms$symbol),]
rownames(tpms) = NULL
tpms = tpms |>  column_to_rownames("symbol") 

# keep mRNA
tpms = tpms[tpms$type == "protein_coding",]
tpms = tpms[,-c(1,ncol(tpms))]
colnames(tpms) = substring(colnames(tpms),1,16)
tpms = tpms[,!duplicated(colnames(tpms))]
tpms01A = tpms[, substring(colnames(tpms), 14, 16) == c("01A")]
tpms11A = tpms[, substring(colnames(tpms), 14, 16) == c("11A")]
```

check whether they have same col name and row name

```{r}
identical(rownames(counts01A), rownames(counts11A))
identical(rownames(tpms01A), rownames(tpms11A))
identical(rownames(counts01A), rownames(tpms01A))
identical(colnames(counts01A), colnames(tpms01A))
identical(colnames(counts11A), colnames(tpms11A))
```

## output data

```{r}
# write.table(
#   counts01A,
#   "./data_file/counts01A.txt",
#   sep = "\t",
#   row.names = T,
#   col.names = NA,
#   quote = F
# )
# write.table(
#   counts11A,
#   "./data_file/counts11A.txt",
#   sep = "\t",
#   row.names = T,
#   col.names = NA,
#   quote = F
# )
# write.table(
#   tpms01A,
#   "./data_file/tpms01A.txt",
#   sep = "\t",
#   row.names = T,
#   col.names = NA,
#   quote = F
# )
# write.table(
#   tpms11A,
#   "./data_file/tpms11A.txt",
#   sep = "\t",
#   row.names = T,
#   col.names = NA,
#   quote = F
# )
```

## combine data

```{r}
counts = cbind(counts01A, counts11A)
tpms = cbind(tpms01A, tpms11A)
# write.table(
#   counts,
#   "./data_file/counts.txt",
#   sep = "\t",
#   row.names = T,
#   col.names = NA,
#   quote = F
# )
# write.table(
#   tpms,
#   "./data_file/tpms.txt",
#   sep = "\t",
#   row.names = T,
#   col.names = NA,
#   quote = F
# )
```

## transform the data

```{r}
range(tpms)
range(tpms01A)
range(tpms11A)
tpms_log2 = log2(tpms + 1)
range(tpms_log2)
tpms01A_log2 = log2(tpms01A + 1)
range(tpms01A_log2)
tpms11A_log2 = log2(tpms11A + 1)
range(tpms11A_log2)

# write.table(
#   tpms_log2,
#   "./data_file/tpms_log2.txt",
#   sep = "\t",
#   row.names = T,
#   col.names = NA,
#   quote = F
# )
# write.table(
#   tpms01A_log2,
#   "./data_file/tpms01A_log2.txt",
#   sep = "\t",
#   row.names = T,
#   col.names = NA,
#   quote = F
# )
# write.table(
#   tpms11A_log2,
#   "./data_file/tpms11A_log2.txt",
#   sep = "\t",
#   row.names = T,
#   col.names = NA,
#   quote = F
# )
```

# survival analysis

```{r}
#计算患者免疫评分与肿瘤纯度# 基质组分 免疫组分 肿瘤组分 肿瘤纯度
library(utils)
# rforge = "http://r-forge.r-project.org"
# install.packages("estimate", repos=rforge, dependencies=TRUE)
library(estimate)
```

```{r}
exp = read.table(
  "./data_file/tpms01A_log2.txt",
  sep = "\t",
  row.names = 1,
  check.names = F,
  stringsAsFactors = F,
  header = T
)
```

## calculate immune score

```{r}
# #计算免疫评分
# filterCommonGenes(input.f = "./data_file/tpms01A_log2.txt",
#                   #输入文件名
#                   output.f = "./data_file/tpms01A_log2.gct",
#                   #输出文件名
#                   id = "GeneSymbol")   #行名为gene symbol
# estimateScore("./data_file/tpms01A_log2.gct",
#               #刚才的输出文件名
#               "./data_file/tpms01A_log2_estimate_score.txt",
#               #新的输出文件名（即估计的结果文件）
#               platform = "affymetrix")   #默认平台
```

extract the outcome and clean data

```{r}
ESTIMATE_result =
  read.table(
    "./data_file/tpms01A_log2_estimate_score.txt",
    sep = "\t",
    row.names = 1,
    check.names = F,
    stringsAsFactors = F,
    header = T
  )
ESTIMATE_result = ESTIMATE_result[,-1]
colnames(ESTIMATE_result) = ESTIMATE_result[1,]
ESTIMATE_result = as.data.frame(t(ESTIMATE_result[-1,]))
rownames(ESTIMATE_result) = colnames(exp)
```

export result

```{r}
# write.table(
#   ESTIMATE_result,
#   file = "./data_file/ESTIMATE_result.txt",
#   sep = "\t",
#   row.names = T,
#   col.names = NA,
#   quote = F
# )
```

## get survival data

```{r survival analysis}
#xena homepage: https://xenabrowser.net/datapages/
# download
survival =
  read.table(
    "./data_file/OS.txt",
    sep = "\t",
    row.names = 1,
    check.names = F,
    stringsAsFactors = F,
    header = T
  ) |>
  _[, 2:3]
survival = survival |>  rownames_to_column('sample')
survival$name = paste0(survival$sample, 'A')#paste粘贴、连接
table(substring(survival$name, 14, 16))
rownames(survival) = survival$name
survival = survival[, 2:3]
```

combine survival data with gene expression

```{r}
tpms01A_log2 =
  read.table(
    "./data_file/tpms01A_log2.txt",
    sep = "\t",
    row.names = 1,
    check.names = F,
    header = T
  )
# get rownames in common
a = intersect(colnames(tpms01A_log2), rownames(survival))
table(substr(a,14,16))
exp_01A = tpms01A_log2[, a]
surv_01A = survival[a, ]
exp_01A = exp_01A |>  t() |>  as.data.frame()
identical(rownames(exp_01A), rownames(surv_01A))
exp_surv_01A = cbind(surv_01A, exp_01A)

# write.table(
#   exp_surv_01A,
#   "./data_file/exp_surv_01A.txt",
#   sep = "\t",
#   row.names = T,
#   col.names = NA,
#   quote = F
# )
```

combine survival data with gene score

```{r}
ESTIMATE_result =
  read.table(
    "./data_file/ESTIMATE_result.txt",
    sep = "\t",
    row.names = 1,
    check.names = F,
    header = T
  )
identical(rownames(ESTIMATE_result), rownames(surv_01A))
ESTIMATE_result_surv_01A = cbind(surv_01A, ESTIMATE_result)

# write.table(
#   ESTIMATE_result_surv_01A,
#   "./data_file/ESTIMATE_result_surv_01A.txt",
#   sep = "\t",
#   row.names = T,
#   col.names = NA,
#   quote = F
# )
```

## generate survival function

```{r}
surv =
  read.table(
    "./data_file/ESTIMATE_result_surv_01A.txt",
    sep = "\t",
    row.names = 1,
    check.names = F,
    stringsAsFactors = F,
    header = T
  )
surv$OS.time = surv$OS.time / 365
```

```{r function}
library(survival)
library(survminer)
surval_fun = function(var, surv_data) {
  surv_data$group =
    ifelse(surv_data[[var]] > median(surv_data[[var]]), "High", "Low")
  class(surv_data$group)
  surv$group = factor(surv_data$group, levels = c("Low", "High"))
  class(surv_data$group)
  table(surv_data$group)
  # cox ph model
  fitd = survdiff(Surv(OS.time, OS) ~ group,
                  data      = surv_data,
                  na.action = na.exclude)
  pValue = 1 - pchisq(fitd$chisq, length(fitd$n) - 1)
  # plot survival curve
  fit = survfit(Surv(OS.time, OS) ~ group, data = surv_data)
  summary(fit)
  p.lab =
    paste0("P", ifelse(pValue < 0.001, " < 0.001", paste0(" = ", round(pValue, 3))))
  ggsurvplot(
    fit,
    data = surv_data,
    pval = p.lab,
    conf.int = TRUE,
    # 显示置信区间
    risk.table = TRUE,
    # 显示风险表
    risk.table.col = "strata",
    palette = "jco",
    # 配色采用jco
    legend.labs = c("Low", "High"),
    # 图例
    size = 1,
    xlim = c(0, 20),
    # x轴长度
    break.time.by = 5,
    # x轴步长为5
    legend.title = var,
    surv.median.line = "hv",
    # 限制垂直和水平的中位生存
    ylab = "Survival probability (%)",
    # 修改y轴标签
    xlab = "Time (Years)",
    # 修改x轴标签
    ncensor.plot = TRUE,
    # 显示删失图块
    ncensor.plot.height = 0.25,
    risk.table.y.text = FALSE
  )
  # dev.off()
}
```

execute function

```{r}
surval_fun(var = "ImmuneScore", surv_data = surv)
surval_fun(var = "StromalScore", surv_data = surv)
surval_fun(var = "ESTIMATEScore", surv_data = surv)
```

# along with tumor stage

## get the tumor stage

we can get the breakdown of the different tumor stages

```{r}
# 随着肿瘤分期的进展
clinical = as.data.frame(expquery2@colData) |>
  distinct(sample, .keep_all = TRUE) |>
  select(
    gender,
    age_at_index,
    ajcc_pathologic_stage,
    ajcc_pathologic_t,
    ajcc_pathologic_n,
    ajcc_pathologic_m
  )

class(clinical$gender)
class(clinical$age_at_index)
class(clinical$ajcc_pathologic_stage)
class(clinical$ajcc_pathologic_t)
class(clinical$ajcc_pathologic_n)
class(clinical$ajcc_pathologic_m)

table(clinical$gender)
table(clinical$age_at_index)
table(clinical$ajcc_pathologic_stage)
table(clinical$ajcc_pathologic_t)
table(clinical$ajcc_pathologic_n)
table(clinical$ajcc_pathologic_m)
```

simplify the stage

```{r}
# gsub = replace
clinical$ajcc_pathologic_stage =
  gsub("A", "", clinical$ajcc_pathologic_stage)
clinical$ajcc_pathologic_stage =
  gsub("B", "", clinical$ajcc_pathologic_stage)
clinical$ajcc_pathologic_t =
  gsub("a", "", clinical$ajcc_pathologic_t)
clinical$ajcc_pathologic_t =
  gsub("b", "", clinical$ajcc_pathologic_t)
clinical$ajcc_pathologic_m =
  gsub("a", "", clinical$ajcc_pathologic_m)
clinical$ajcc_pathologic_m =
  gsub("b", "", clinical$ajcc_pathologic_m)

#提取01A临床数据
rownames(clinical) = substring(rownames(clinical), 1, 16)
```

combine the clinical data with gene expression

```{r}
##将基因表达谱和临床数据合并并保存
exp01A =
  read.table(
    "./data_file/tpms01A_log2.txt",
    sep = "\t",
    row.names = 1,
    check.names = F,
    stringsAsFactors = F,
    header = T
  )

clinical01A = clinical[colnames(exp01A),]
exp01A = exp01A |>  t() |>  as.data.frame()
identical(rownames(clinical01A),rownames(exp01A))
clinical.expr01A = cbind(clinical01A,exp01A)

# write.table(
#   clinical.expr01A,
#   "./data_file/clinical.expr01A.txt",
#   sep = "\t",
#   row.names = T,
#   col.names = NA,
#   quote = F
# )

ESTIMATE_result =
  read.table(
    "./data_file/ESTIMATE_result.txt",
    sep = "\t",
    row.names = 1,
    check.names = F,
    stringsAsFactors = F,
    header = T
  )
identical(rownames(clinical01A), rownames(ESTIMATE_result))
clinical.ESTIMATE_result01A = cbind(clinical01A, ESTIMATE_result)

# write.csv(clinical.ESTIMATE_result01A, file = "./data_file/clinical.ESTIMATE_result01A.csv")
```

```{r}
clinical.ESTIMATE_result01A |>
  filter(!is.na(ajcc_pathologic_stage)) |> 
  ggplot(aes(x = ajcc_pathologic_stage,
             y = ImmuneScore,
             fill = ajcc_pathologic_stage)) +
  geom_boxplot() +
  labs(title = "Stage",
       y = "Immune Score") +
  theme_minimal() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 14, face = "bold"),
    legend.position = NULL
  )
```

calculate p-value

```{r}
test = 
clinical.ESTIMATE_result01A |>
  filter(!is.na(ajcc_pathologic_stage)) |>
  mutate(ajcc_pathologic_stage = as.factor(ajcc_pathologic_stage)) %>%  kruskal.test(ImmuneScore ~ ajcc_pathologic_stage)
  
kruskal.test(test$ImmuneScore ~ test$ajcc_pathologic_stage)
```

# difference analysis

## import data

```{r}
# BiocManager::install('DESeq2')
library(DESeq2)
counts_01A =
  read.table(
    "./data_file/counts01A.txt",
    sep = "\t",
    row.names = 1,
    check.names = F,
    stringsAsFactors = F,
    header = T
  )
estimate =
  read.table(
    "./data_file/ESTIMATE_result.txt",
    sep = "\t",
    row.names = 1,
    check.names = F,
    header = T
  )
```

## ImmuneScore difference score

```{r}
x = "ImmuneScore"
med = as.numeric(median(estimate[, x]))
estimate = as.data.frame(t(estimate))
if (identical(colnames(counts_01A), colnames(estimate)))
{
  conditions = data.frame(sample = colnames(counts_01A),
                          group = factor(ifelse(estimate[x, ] > med, "high", "low"), levels = c("low", "high"))) |> column_to_rownames("sample")
} else {
  cat("colnames are different")
}
# prepare to make a difference analysis
dds = DESeqDataSetFromMatrix(countData = counts_01A,
                             colData = conditions,
                             design = ~ group)
# start to make a difference analysis
dds = DESeq(dds)
resultsNames(dds)

# get result
res = results(dds)
# save(res, file = "./data_file/DEG_ImmuneScore.Rda")
```

plot heatmap

```{r}
DEG = as.data.frame(res)
# read expression spectrum (tpms)
exp = read.table(
  "./data_file/tpms01A_log2.txt",
  sep = "\t",
  row.names = 1,
  check.names = F,
  stringsAsFactors = F,
  header = T
)
# define the cutoff value
logFC_cutoff = 1
type1 = (DEG$padj < 0.05)&(DEG$log2FoldChange < -logFC_cutoff)
type2 = (DEG$padj < 0.05)&(DEG$log2FoldChange > logFC_cutoff)
DEG$change = ifelse(type1,"DOWN",ifelse(type2,"UP","NOT"))
table(DEG$change)

# install.packages("pheatmap")
library(pheatmap)
# extract different genes' expression spectrum
a = filter(DEG, change == 'UP')
b = filter(DEG, change == 'DOWN')
c = rbind(a, b)
d = rownames(c)
exp_diff = exp[d, ]

# set group information
annotation_col = conditions
# adjust the sequence of sample
a = filter(annotation_col, group == 'high')
b = filter(annotation_col, group == 'low')
exp_diff_high = exp_diff[, rownames(a)]
exp_diff_low = exp_diff[, rownames(b)]
exp_diff = cbind(exp_diff_high, exp_diff_low)

# plot heatmap
pheatmap(
  exp_diff,
  annotation_col = annotation_col,
  scale = "row",
  show_rownames = F,
  show_colnames = F,
  color = colorRampPalette(c("navy", "white", "red"))(50),
  cluster_cols = F,
  cluster_rows = T,
  fontsize = 10,
  fontsize_row = 3,
  fontsize_col = 3
)
# dev.off()
```

## StromalScore difference score

```{r}
counts_01A =
  read.table(
    "./data_file/counts01A.txt",
    sep = "\t",
    row.names = 1,
    check.names = F,
    stringsAsFactors = F,
    header = T
  )
estimate =
  read.table(
    "./data_file/ESTIMATE_result.txt",
    sep = "\t",
    row.names = 1,
    check.names = F,
    header = T
  )
```

```{r}
x = "StromalScore"
med = as.numeric(median(estimate[, x]))
estimate = as.data.frame(t(estimate))
if (identical(colnames(counts_01A), colnames(estimate)))
{
  conditions = data.frame(sample = colnames(counts_01A),
                          group = factor(ifelse(estimate[x, ] > med, "high", "low"), levels = c("low", "high"))) |> column_to_rownames("sample")
} else {
  cat("colnames are different")
}
# prepare to make a difference analysis
dds = DESeqDataSetFromMatrix(countData = counts_01A,
                             colData = conditions,
                             design = ~ group)
# start to make a difference analysis
dds = DESeq(dds)
resultsNames(dds)

# get result
res = results(dds)
# save(res, file = "./data_file/DEG_StromalScore.Rda")
```

plot heatmap

```{r}
DEG = as.data.frame(res)
# read expression spectrum (tpms)
exp = read.table(
  "./data_file/tpms01A_log2.txt",
  sep = "\t",
  row.names = 1,
  check.names = F,
  stringsAsFactors = F,
  header = T
)
# define the cutoff value
logFC_cutoff = 1
type1 = (DEG$padj < 0.05)&(DEG$log2FoldChange < -logFC_cutoff)
type2 = (DEG$padj < 0.05)&(DEG$log2FoldChange > logFC_cutoff)
DEG$change = ifelse(type1,"DOWN",ifelse(type2,"UP","NOT"))
table(DEG$change)

# extract different genes' expression spectrum
a = filter(DEG, change == 'UP')
b = filter(DEG, change == 'DOWN')
c = rbind(a, b)
d = rownames(c)
exp_diff = exp[d, ]

# set group information
annotation_col = conditions
# adjust the sequence of sample
a = filter(annotation_col, group == 'high')
b = filter(annotation_col, group == 'low')
exp_diff_high = exp_diff[, rownames(a)]
exp_diff_low = exp_diff[, rownames(b)]
exp_diff = cbind(exp_diff_high, exp_diff_low)

# plot heatmap
pheatmap(
  exp_diff,
  annotation_col = annotation_col,
  scale = "row",
  show_rownames = F,
  show_colnames = F,
  color = colorRampPalette(c("navy", "white", "red"))(50),
  cluster_cols = F,
  cluster_rows = T,
  fontsize = 10,
  fontsize_row = 3,
  fontsize_col = 3
)
# dev.off()
```

## Get the intersect gene

```{r}
load("./data_file/DEG_ImmuneScore.Rda")
DEG = as.data.frame(res)
logFC_cutoff = 1
type1 = (DEG$padj < 0.05) & (DEG$log2FoldChange < -logFC_cutoff)
type2 = (DEG$padj < 0.05) & (DEG$log2FoldChange > logFC_cutoff)
DEG$change = ifelse(type1, "DOWN", ifelse(type2, "UP", "NOT"))
table(DEG$change)
DEG_IMM_UP = filter(DEG, change == 'UP')
DEG_IMM_DOWN = filter(DEG, change == 'DOWN')

load("./data_file/DEG_StromalScore.Rda")
DEG = as.data.frame(res)
logFC_cutoff = 1
type1 = (DEG$padj < 0.05) & (DEG$log2FoldChange < -logFC_cutoff)
type2 = (DEG$padj < 0.05) & (DEG$log2FoldChange > logFC_cutoff)
DEG$change = ifelse(type1, "DOWN", ifelse(type2, "UP", "NOT"))
table(DEG$change)
DEG_STR_UP = filter(DEG, change == 'UP')
DEG_STR_DOWN = filter(DEG, change == 'DOWN')
```

plot the Venn Diagram

```{r}
# install.packages("VennDiagram")
library(VennDiagram)
library(ggsci)
jco_colors = pal_jco("default")(2)

venn.plot_up = draw.pairwise.venn(
  area1 = length(rownames(DEG_IMM_UP)),
  area2 = length(rownames(DEG_STR_UP)),
  cross.area = length(intersect(
    rownames(DEG_IMM_UP), rownames(DEG_STR_UP)
  )),
  category = c("ImmuneScore up", "StromalScore up"),
  fill = jco_colors,
  cat.cex = 1.2,  
  cat.col = jco_colors,  
  lty = "blank"  
)

venn.plot_down = draw.pairwise.venn(
  area1 = length(rownames(DEG_IMM_DOWN)),
  area2 = length(rownames(DEG_STR_DOWN)),
  cross.area = length(intersect(
    rownames(DEG_IMM_DOWN), rownames(DEG_STR_DOWN)
  )),
  category = c("ImmuneScore down", "StromalScore down"),
  fill = jco_colors,
  cat.cex = 1.2,  
  cat.col = jco_colors,  
  lty = "blank"  
)
```

## make an Enrichment Analysis with intersection gene

```{r}
# BiocManager::install('clusterProfiler')
# BiocManager::install('org.Hs.eg.db')
library(org.Hs.eg.db)
#org.Hs.eg.db包主要注释人类基因:用于不同数据库ID间的转化
library(clusterProfiler)

# use immunescore or stromalscore
load("./data_file/DEG_ImmuneScore.Rda")
DEG = as.data.frame(res)
DEG_final_up = as.data.frame(intersect(rownames(DEG_IMM_UP), rownames(DEG_STR_UP)))
DEG_final_down = as.data.frame(intersect(rownames(DEG_IMM_DOWN), rownames(DEG_STR_DOWN)))
colnames(DEG_final_up) = c("SYMBOL")
colnames(DEG_final_down) = c("SYMBOL")

DEG_final = rbind(DEG_final_up, DEG_final_down)
# write.table(
#   DEG_final,
#   file = "./data_file/DEG_final.txt",
#   sep = "\t",
#   row.names = FALSE,
#   quote = FALSE
# )

DEG = DEG[DEG_final$SYMBOL,]
DEG = rownames_to_column(DEG, "SYMBOL")

# transform id
genelist = bitr(DEG$SYMBOL,
                fromType = "SYMBOL",
                toType = "ENTREZID",
                OrgDb = 'org.Hs.eg.db')
DEG = inner_join(DEG, genelist, by = "SYMBOL")
```

GO function

```{r}
####GO####
ego = enrichGO(
  gene = DEG$ENTREZID,
  OrgDb = org.Hs.eg.db,
  # get all functions
  ont = "all",
  pAdjustMethod = "BH",
  minGSSize = 1,
  pvalueCutoff = 0.05,
  qvalueCutoff = 0.05,
  readable = TRUE
)

ego_res = ego@result
# save(ego, ego_res, file = "./data_file/GO_DEG_final.Rda")
```

KEGG function

```{r}
####KEGG####
kk = enrichKEGG(
  gene         = DEG$ENTREZID,
  organism     = 'hsa',
  pvalueCutoff = 0.1,
  qvalueCutoff = 0.1
)
kk_res = kk@result
# save(kk, kk_res, file = "./data_file/KEGG_DEG_final.Rda")
```

## plot the gride

```{r}
# install.packages("ggnewscale")
library(ggnewscale)
List = DEG$log2FoldChange
names(List)= DEG$ENTREZID
head(List)
List = sort(List, decreasing = T)

#GO
cnetplot(ego,
         foldChange = List,
         circular = TRUE,
         colorEdge = TRUE)
#KEGG
cnetplot(kk,
         foldChange = List,
         circular = TRUE,
         colorEdge = TRUE)
```

## PPI (protein and protein interaction)

```{r}
# String：https://cn.string-db.org/cgi/input?sessionId=bXmYsv7CnUrH&input_page_active_form=multiple_identifiers
# download the TSV file and use a Cytoscape to plot 

data_PPI =
  read_csv("./data_file/string_interactions_short_tsv_default_node.csv")

data_PPI =
  data_PPI |> select(name, Degree) |>
  janitor::clean_names() |>
  arrange(desc(degree))

# plot the most influential protein mRNA and the # of their connections

data_PPI[1:30,] |>
  ggplot(aes(x = fct_reorder(name, degree), y = degree)) + geom_col() +
  coord_flip() + 
  theme_minimal() +
  labs(title = "Degree Distribution",
       x = "Name",
       y = "Degree") +
  theme(axis.text.x = element_text(angle = 0, hjust = 1))
```

## COX with intersection different genes

```{r}
library(survival)
# install.packages("forestplot")
library(forestplot)
exp_surv_01A = read.table(
  "./data_file/exp_surv_01A.txt",
  sep = "\t",
  row.names = 1,
  check.names = F,
  stringsAsFactors = F,
  header = T
)

surv.expr = cbind(exp_surv_01A[, 1:2], exp_surv_01A[, DEG_final$SYMBOL])
# colnames(surv.expr)[1] = "censor" 
```

```{r}
Coxoutput = NULL

for (i in 3:ncol(surv.expr)) {
  g = colnames(surv.expr)[i]
  cox =
    coxph(Surv(OS.time, OS) ~ surv.expr[, i], data = surv.expr) # univariate model
  coxSummary = summary(cox)
  
  Coxoutput = rbind.data.frame(
    Coxoutput,
    data.frame(
      gene = g,
      HR = as.numeric(coxSummary$coefficients[, "exp(coef)"])[1],
      z = as.numeric(coxSummary$coefficients[, "z"])[1],
      pvalue = as.numeric(coxSummary$coefficients[, "Pr(>|z|)"])[1],
      lower = as.numeric(coxSummary$conf.int[, 3][1]),
      upper = as.numeric(coxSummary$conf.int[, 4][1]),
      stringsAsFactors = F
    ),
    stringsAsFactors = F
  )
}

Coxoutput = arrange(Coxoutput, pvalue)

# select the top genes
gene_sig = Coxoutput[Coxoutput$pvalue < 0.001, ] # if 0.05, there are too many genes to consider
# write.csv(gene_sig, file = "./data_file/gene_sig.csv")
topgene = gene_sig 
```

plot the forest plot

```{r}
# make the form first
tabletext = cbind(
  c("Gene", topgene$gene),
  c("HR", format(round(
    as.numeric(topgene$HR), 3
  ), nsmall = 3)),
  c("lower 95%CI", format(round(
    as.numeric(topgene$lower), 3
  ), nsmall = 3)),
  c("upper 95%CI", format(round(
    as.numeric(topgene$upper), 3
  ), nsmall = 3)),
  c("pvalue", format(round(
    as.numeric(topgene$p), 3
  ), nsmall = 3))
)
```

```{r }
# plot forest
forestplot(
  labeltext = tabletext,
  mean = c(NA, as.numeric(topgene$HR)),
  lower = c(NA, as.numeric(topgene$lower)),
  upper = c(NA, as.numeric(topgene$upper)),
  graph.pos = 6,
  # 图在表中的列位置
  graphwidth = unit(.75, "npc"),
  # 图在表中的宽度比
  fn.ci_norm = "fpDrawDiamondCI",
  # box类型选择钻石
  col = fpColors(
    box = "#00A896",
    lines = "#02C39A",
    zero = "black"
  ),
  # box颜色
  boxsize = 0.3,
  # box大小固定
  lwd.ci = 1,
  ci.vertices.height = 0.1,
  ci.vertices = T,
  # 显示区间
  zero = 1,
  # zero线横坐标
  lwd.zero = 1.5,
  # zero线宽
  xticks = c(0.5, 1, 1.5),
  # 横坐标刻度根据需要可随意设置
  lwd.xaxis = 2,
  xlab = "Hazard ratios",
  txt_gp = fpTxtGp(
    label = gpar(cex = 1.2),
    # 各种字体大小设置
    ticks = gpar(cex = 0.85),
    xlab = gpar(cex = 1),
    title = gpar(cex = 1.5)
  ),
  hrzl_lines = list(
    "1" = gpar(lwd = 2, col = "black"),
    # 在第一行上面画黑色实线
    "2" = gpar(lwd = 1.5, col = "black"),
    # 在第一行标题行下画黑色实线
    "22" = gpar(lwd = 2, col = "black")
  ),
  # 在最后一行上画黑色实线
  lineheight = unit(.5, "cm"),
  # 固定行高
  colgap = unit(0.3, "cm"),
  mar = unit(rep(1.5, times = 4), "cm"),
  new_page = F
)
# dev.off()
```

select the intersection gene in protein interaction and survival analysis

```{r}
gene_sig = Coxoutput[Coxoutput$pvalue < 0.005, ]

jco_colors = pal_jco("default")(2)

venn.plot_up = draw.pairwise.venn(
  area1 = length(gene_sig$gene),
  area2 = length(data_PPI[1:30,]$name),
  cross.area = length(intersect(gene_sig$gene, data_PPI[1:30,]$name)),
  category = c("COX", "PPI network"),
  fill = jco_colors,
  cat.cex = 1.2,  
  cat.col = jco_colors,  
  lty = "blank"  
)
intersect(gene_sig$gene, data_PPI[1:30,]$name)
```

We select "BTK" gene to make a further study

# gene BTK expression difference in cancer and no cancer samples

boxplot in different samples

```{r}
tpms01A_log2 = read.table(
  "./data_file/tpms01A_log2.txt",
  sep = "\t",
  row.names = 1,
  check.names = F,
  stringsAsFactors = F,
  header = T
)
tpms11A_log2 = read.table(
  "./data_file/tpms11A_log2.txt",
  sep = "\t",
  row.names = 1,
  check.names = F,
  stringsAsFactors = F,
  header = T
)
gene = "BTK"
a = tpms01A_log2[gene, ]
b = tpms11A_log2[gene, ]
a = a |>  t() |>  as.data.frame()
b = b |>  t() |>  as.data.frame()
# write.csv(a, file = "./data_file/BTK_01A.csv")
# write.csv(b, file = "./data_file/BTK_11A.csv")
a = a |>
  mutate(group = "tumor")
rownames(a) = NULL

b = b |>
  mutate(group = "normal")
rownames(b) = NULL
final = rbind(b, a)

final |> ggplot(aes(x = group, y = BTK)) + 
  geom_boxplot() + 
  theme_classic() + 
  labs(title = "BTK expression",
       y = "BTK")
```

pair plot in common sample

```{r}
tpms01A_log2 = tpms01A_log2 |>  t() |>  as.data.frame()
tpms11A_log2 = tpms11A_log2 |>  t() |>  as.data.frame()
rownames(tpms01A_log2) = substring(rownames(tpms01A_log2), 1, 12)
rownames(tpms11A_log2) = substring(rownames(tpms11A_log2), 1, 12)
a = intersect(rownames(tpms01A_log2),rownames(tpms11A_log2))
tpms01A_log2 = tpms01A_log2[a,]
tpms11A_log2 = tpms11A_log2[a,]
peidui = cbind(tpms11A_log2[, gene], tpms01A_log2[, gene])#11A放在前面
peidui = as.data.frame(peidui)
colnames(peidui) = c("normal", "tumor")
# write.csv(peidui,file = "./data_file/peidui.csv")
peidui =
  peidui |> 
  mutate(patient = row_number()) |> 
  pivot_longer(cols = c("normal","tumor"), 
               names_to = "group", 
               values_to = "value") |> 
  arrange(group)
```

```{r}
library(ggpubr)
ggplot(peidui,
       aes(
         x = group,
         y = value,
         group = patient,
         color = group
       )) +
  geom_point(size = 3) +
  geom_line(aes(group = patient), color = "black") +
  scale_color_manual(values = c("normal" = "blue", "tumor" = "orange")) +
  theme_classic() +
  labs(title = "Paired Comparison: Normal vs Tumor", y = "BTK Value") +
  theme(legend.position = "none") +
  stat_compare_means(method = "t.test", aes(group = group), label = "p.signif")
```

make an survival analysis according to different BTK expression

```{r}
# dichotomize the BTK expression
surv = read.table(
  "./data_file/exp_surv_01A.txt",
  sep = "\t",
  row.names = 1,
  check.names = F,
  stringsAsFactors = F,
  header = T
)
surv$OS.time = surv$OS.time / 365
```

```{r}
library(survival)
#median 
#BTK
surv$group = ifelse(surv$BTK > median(surv$BTK), "High", "Low")
class(surv$group)
surv$group = factor(surv$group, levels = c("Low", "High")) 
class(surv$group)
table(surv$group)
fitd = survdiff(Surv(OS.time, OS) ~ group,
                data      = surv,
                na.action = na.exclude)
pValue = 1 - pchisq(fitd$chisq, length(fitd$n) - 1)
```

# plot survival curve

```{r}
fit = survfit(Surv(OS.time, OS) ~ group, data = surv)
summary(fit)
p.lab = paste0("P", ifelse(pValue < 0.001, " < 0.001", paste0(" = ", round(pValue, 3))))

library(survminer)
ggsurvplot(
  fit,
  data = surv,
  pval = p.lab,
  conf.int = TRUE,
  # 显示置信区间
  risk.table = TRUE,
  # 显示风险表
  risk.table.col = "strata",
  palette = "jco",
  # 配色采用jco
  legend.labs = c("Low", "High"),
  # 图例
  size = 1,
  xlim = c(0, 20),
  # x轴长度
  break.time.by = 5,
  # x轴步长为5
  legend.title = "BTK",
  surv.median.line = "hv",
  # 限制垂直和水平的中位生存
  ylab = "Survival probability (%)",
  # 修改y轴标签
  xlab = "Time (Years)",
  # 修改x轴标签
  ncensor.plot = TRUE,
  # 显示删失图块
  ncensor.plot.height = 0.25,
  risk.table.y.text = FALSE
)
```

## different expression of BTK in different stages

```{r}
clinical.expr01A = read.table(
  "./data_file/clinical.expr01A.txt",
  sep = "\t",
  row.names = 1,
  check.names = F,
  stringsAsFactors = F,
  header = T
)
gene = "BTK"
clinical_BTK =
  cbind(clinical.expr01A[, 1:6], clinical.expr01A[, gene])
# write.csv(clinical_BTK, file = "./data_file/clinical_BTK.csv")
```

```{r}

```

# make a difference analysis with BTK

```{r}
library(DESeq2)
counts_01A = read.table(
  "./data_file/counts01A.txt",
  sep = "\t",
  row.names = 1,
  check.names = F,
  stringsAsFactors = F,
  header = T
)
exp = read.table(
  "./data_file/tpms01A_log2.txt",
  sep = "\t",
  row.names = 1,
  check.names = F,
  header = T
)
identical(colnames(counts_01A),colnames(exp))
gene = "BTK"
med = median(as.numeric(exp[gene, ]))

conditions =
  data.frame(sample = colnames(exp),
             group = factor(ifelse(exp[gene,] > med, "high", "low"), levels = c("low", "high"))) |>
  column_to_rownames("sample")

dds = DESeqDataSetFromMatrix(countData = counts_01A,
                             colData = conditions,
                             design = ~ group)
dds = DESeq(dds)
resultsNames(dds)

res = results(dds)
save(res,file="./data_file/DEG_BTK.Rda")
```

